src/ruby/puppetserver-lib/puppet/server/config.rb

```diff
@@ -42,17 +46,79 @@ def self.initialize_puppet_server(puppet_server_config)
     end
   end
 
-  def self.ssl_context
+  def self.puppet_only_ssl_context
     # Initialize an SSLContext for use during HTTPS client requests.
     # Do this lazily due to startup-ordering issues - to give the CA
     # service time to create these files before they are referenced here.
-    unless @ssl_context
-      @ssl_context = SSLUtils.pems_to_ssl_context(
+    unless @puppet_only_ssl_context
+      @puppet_only_ssl_context = SSLUtils.pems_to_ssl_context(
           FileReader.new(Puppet[:hostcert]),
           FileReader.new(Puppet[:hostprivkey]),
           FileReader.new(Puppet[:localcacert]))
     end
-    @ssl_context
+    @puppet_only_ssl_context
+  end
+
+  def self.ssl_context
+    self.puppet_only_ssl_context
+  end
+
+  # The logging for this method strives to match the equivalent
+  # agent logging in 'puppet/ssl/ssl_provider'
+  def self.puppet_and_system_ssl_context
+    unless @puppet_and_system_ssl_context
+      @puppet_and_system_ssl_context = load_puppet_and_system_ssl_context
+    end
+
+    @puppet_and_system_ssl_context
+  end
+
+  def self.load_puppet_and_system_ssl_context
+    stores = SSLUtils.pemsToKeyAndTrustStores(
+        FileReader.new(Puppet[:hostcert]),
+        FileReader.new(Puppet[:hostprivkey]),
+        FileReader.new(Puppet[:localcacert]))
+
+    truststore = stores['truststore']
+    if File.exist?(PUPPET_KEYSTORE_LOCATION)
+      associate_entries(truststore, PUPPET_KEYSTORE_LOCATION)
+    end
+
+    if additional_store_location = Puppet[:ssl_trust_store]
+      if File.exist?(additional_store_location)
+        begin
+          if File.read(additional_store_location) =~ CERT_REGEX
+            # The truststore looks like a cert chain the agent reads
+            # Attempt to load it as SSLUtils would our own CA chain
+            SSLUtils.associateCertsFromReader(
+              truststore,
+              'puppet_setting_ssl_trust_store',
+              FileReader.new(additional_store_location))
+
+          else
+            # Attempt to treat the file as a java keystore
+            associate_entries(truststore, additional_store_location)
+          end
+        rescue => detail
+          Puppet.err("Failed to add '#{additional_store_location}' as a trusted CA file: #{detail}")
+        end
+      else
+        Puppet.warning("The 'ssl_trust_store' setting does not refer to a file and will be ignored: '#{additional_store_location}'")
```

Comment:
Yeah, I went back and forth on the readability of having a branch dangle so far from the condition's test vs sandwiching the main logic between large warnings and swapping positive vs negative conditionals. What if kept the structure but pulled out another helper method so the main branch was just a line? eg:
```rb
    if additional_store_location = Puppet[:ssl_trust_store]
      if File.exist?(additional_store_location)
        begin
          load_additional_store(additional_store_location)
        rescue => detail
          Puppet.err("Failed to add '#{additional_store_location}' as a trusted CA file: #{detail}")
        end
      else
        Puppet.warning("The 'ssl_trust_store' setting does not refer to a file and will be ignored: '#{additional_store_location}'")
    end
```
mmm... in that case maybe it would be better to pull the rescue into that new helper method as well. What do you think?

Reply:
I think a `load_additional_store` method would help the readability. and yeah, if you add that, I think the `rescue` should definitely go in there.